src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,293 @@
+          fakeData.n_rows, batchSize, false, false);
+      tempTrainData.cols(batchSize, tempTrainData.n_cols - 1) = arma::mat(
+          predictors.memptr() + offset, predictors.n_rows, batchSize,
+          false, false);
+      tempLabels.cols(0, batchSize - 1) = arma::mat(
+          fakeLables.memptr() + offset, fakeLables.n_rows, batchSize,
+          false, false);
+      tempLabels.cols(batchSize, tempLabels.n_cols - 1) = arma::mat(
+          responses.memptr() + offset, responses.n_rows, batchSize,
+          false, false);
+      offset += batchSize * k;
+      // Training discriminator
+      this->predictors = std::move(tempTrainData);
+      this->responses = std::move(tempLabels);
+      numFunctions = predictors.n_cols;
+      // set here discrminator evaluate function
+      discriminator.predictors = predictors;
+      discriminator.responses = responses;
+      // do one pass over the dataset
+      Optimizer.MaxIterations() = numFunctions;
+      Optimizer.Optimize(*this, parameter);
+    }
+
+    trainGenerator = true;
+    // Training Generator
+    Generate(batchSize, std::move(fakeData), std::move(noiseData));
+    fakeLables.zeros(1, batchSize);
+    this->predictors = std::move(fakeData);
+    this->responses = std::move(fakeLables);
+    numFunctions = predictors.n_cols;
+    discriminator.predictors = predictors;
+    discriminator.responses = responses;
+    Optimizer.MaxIterations() = numFunctions;
+    Optimizer.Optimize(*this, parameter);
+
+    // set train generator to false
+    trainGenerator = false;
+  }
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+double GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Evaluate(const arma::mat& /*parameters*/,
+    const size_t i, const bool /*deterministic*/)
+{
+  currentInput = discriminator.predictors.unsafe_col(i);
+  currentTarget = discriminator.responses.unsafe_col(i);
+  discriminator.Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(std::move(boost::apply_visitor(
+      outputParameterVisitor, discriminator.network.back())),
+      std::move(currentTarget));
+  return res;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Gradient(const arma::mat& parameters, const size_t i, arma::mat& gradient)
+{
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(generator.Parameters().n_rows +
+        discriminator.Parameters().n_rows, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  // Gradient for generator network
+  gradientGenerator = arma::mat(gradient.memptr(),
+      generator.Parameters().n_rows,
+      generator.Parameters().n_cols, false, false);
+
+  // Gradient for discriminator network
+  gradientDisriminator = arma::mat(gradient.memptr(),
```

Comment:
Yes that's what i wanted to do. Sorry i changed this on my local branch did not push it.